core/src/main/scala/kafka/log/LogCleanerManager.scala

```diff
@@ -217,19 +233,44 @@ private[log] class LogCleanerManager(val logDirs: Array[File], val logs: Pool[To
   def updateCheckpoints(dataDir: File, update: Option[(TopicPartition,Long)]) {
     inLock(lock) {
       val checkpoint = checkpoints(dataDir)
-      val existing = checkpoint.read().filterKeys(logs.keys) ++ update
-      checkpoint.write(existing)
+      if (checkpoint != null) {
+        try {
+          val existing = checkpoint.read().filterKeys(logs.keys) ++ update
+          checkpoint.write(existing)
+        } catch {
+          case e: IOException =>
+            error(s"Failed to access checkpoint file ${checkpoint.f}", e)
+            handleLogDirFailure(checkpoint.f.getParentFile.getAbsolutePath)
+        }
+      }
+    }
+  }
+
+  def handleLogDirFailure(dir: String) {
+    info(s"Stopping cleaning logs in dir $dir")
+    inLock(lock) {
+      checkpoints = checkpoints.filterKeys(_.getAbsolutePath != dir)
```

Comment:
If a disk fail in the code path outside of the log cleaner, we should remove that dir from checkpoints too. Is that logic added already?